File: src/go/ch01/1.3/l1/fractions_utils.go
package main

type Fraction struct {
	Num int64
	Den int64
}

// ReduceFraction returns (num, den) in lowest terms with a canonical sign
// (denominator always positive). Panics if d == 0 to keep API simple here.
func ReduceFraction(n, d int64) Fraction {
	if d == 0 {
		panic("denominator cannot be zero")
	}
	g := GCD(n, d)
	num := n / g
	den := d / g
	if den < 0 {
		num, den = -num, -den
	}
	return Fraction{Num: num, Den: den}
}

File: src/python/ch01/1.3/l1/fractions_utils.py
from gcd_lcm import gcd

def reduce_fraction(n: int, d: int) -> tuple[int, int]:
    """Return (num, den) in lowest terms. Raises on zero denominator."""
    if d == 0:
        raise ZeroDivisionError("denominator cannot be zero")
    g = gcd(n, d)
    num = n // g
    den = d // g
    # keep a canonical sign: denominator positive
    if den < 0:
        num, den = -num, -den
    return num, den
